docker/archive/image.go

```diff
@@ -0,0 +1,258 @@
+// Package archive implements handling of `docker load`-style image files. The
+// main purpose of this package is to abstract the tar handling code from
+// docker-daemon:// so that it can be used to also handle local files.
+//
+// In addition, this package provides a docker-archive:// transport which
+// doesn't require access to a Docker daemon.
+package archive
+
+import (
+	"archive/tar"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/containers/image/manifest"
+)
+
+// imageMode represents the "mode" that an Image has been opened in.
+type imageMode int
+
+const (
+	modeSource imageMode = iota
+	modeDestination
+)
+
+var (
+	// ErrReadOnly is returned when an operation is only permitted for images
+	// opened in "read-only" mode (namely ImageSource mode).
+	ErrReadOnly = errors.New("operation not permitted: image is not open in read-only mode")
+
+	// ErrWriteOnly is returned when an operation is only permitted for images
+	// opened in "write-only" mode (namely ImageDestination mode).
+	ErrWriteOnly = errors.New("operation not permitted: image is not open in write-only mode")
+)
+
+// Image implements all of the necessary fields for a Docker image source or
+// image destination with the given path. It should be noted that depending on
+// whether or not the image was open for reading or writing certain operations
+// may not be permitted.
+type Image struct {
+	mode imageMode
+
+	// Only used when open in write mode.
+	// FIXME: Currently we need to use this in order to write out the entire
+	//        archive (if you change tar.Writers it breaks the header setup).
+	//        Is there a nicer way of structuring this?
+	tw *tar.Writer
+
+	// Only used when open in read mode.
+	// FIXME: This caching is ugly. The only reason we have it here is because
+	//        the Docker image format isn't a CAS (you need to store a lookup
+	//        table for diffID -> layerPath).
+	cache *tarCache
```

Comment:
… and then `tarCache` would not be a separate object but just some members of the `TarSource`.